File: app/src/main/java/com/smile/retrofit_package/RestApi.java
package com.smile.retrofit_package;

import android.util.Log;

import com.smile.model.*;

import java.io.IOException;

import retrofit2.Call;
import retrofit2.Retrofit;

public class RestApi {
    private static final String TAG = "RestApi";
    // implement Retrofit to get results synchronously
    public static LanguageList getAllLanguages() {
        Retrofit localRetrofit = Client.getInstance();
        ApiInterface apiInterface = localRetrofit.create(ApiInterface.class);
        Call<LanguageList> call = apiInterface.getAllLanguages();

        LanguageList languageList = null;
        try {
            languageList = call.execute().body();
        } catch (IOException ex) {
            ex.printStackTrace();
        }

        return languageList;
    }

    public static SongList getSongsByLanguage(Language language, int pageSize, int pageNo, String filter) {
        if (language == null) {
            // singer cannot be null
            Log.d(TAG, "getSongsByLanguage.songsList is null.");
            return null;
        }

        int languageId = language.getId();
        String orderBy = "NumWordsSongNa";  // order by (number of words + song's name)

        Retrofit localRetrofit = Client.getInstance();
        ApiInterface apiInterface = localRetrofit.create(ApiInterface.class);
        Call<SongList> call;
        if ( (filter == null) || (filter.isEmpty()) ) {
            call = apiInterface.getSongsByLanguageIdOrderBy(languageId, pageSize, pageNo, orderBy);
        } else {
            call = apiInterface.getSongsByLanguageIdOrderByWithFilter(languageId, pageSize, pageNo, orderBy, filter);
        }

        SongList songList = null;
        try {
            songList = call.execute().body();
        } catch (IOException ex) {
            ex.printStackTrace();
        }

        return songList;
    }

    public static SongList getSongsByLanguageNumOfWords(Language language, int numOfWords, int pageSize, int pageNo, String filter) {
        if (language == null) {
            // singer cannot be null
            Log.d(TAG, "getSongsByLanguageNumOfWords.songsList is null.");
            return null;
        }

        int languageId = language.getId();
        String orderBy = "NumWordsSongNa";  // order by (number of words + song's name)

        Retrofit localRetrofit = Client.getInstance();
        ApiInterface apiInterface = localRetrofit.create(ApiInterface.class);
        Call<SongList> call;
        if ( (filter == null) || (filter.isEmpty()) ) {
            call = apiInterface.getSongsByLanguageIdNumOfWords(languageId, numOfWords, pageSize, pageNo, orderBy);
        } else {
            call = apiInterface.getSongsByLanguageIdNumOfWordsWithFilter(languageId, numOfWords, pageSize, pageNo, orderBy, filter);
        }

        SongList songList = null;
        try {
            songList = call.execute().body();
        } catch (IOException ex) {
            ex.printStackTrace();
        }

        return songList;
    }

    public static SongList getNewSongsByLanguage(Language language, int pageSize, int pageNo, String filter) {
        if (language == null) {
            // singer cannot be null
            Log.d(TAG, "getNewSongsByLanguage.songsList is null.");
            return null;
        }

        int languageId = language.getId();
        String orderBy = "";  // no order. Only the date that the song came in by descending order

        Retrofit localRetrofit = Client.getInstance();
        ApiInterface apiInterface = localRetrofit.create(ApiInterface.class);
        Call<SongList> call;
        if ( (filter == null) || (filter.isEmpty()) ) {
            call = apiInterface.getNewSongsByLanguageId(languageId, pageSize, pageNo);
        } else {
            call = apiInterface.getNewSongsByLanguageIdWithFilter(languageId, pageSize, pageNo, filter);
        }

        SongList songList = null;
        try {
            songList = call.execute().body();
        } catch (IOException ex) {
            ex.printStackTrace();
        }

        return songList;
    }

    public static SongList getHotSongsByLanguage(Language language, int pageSize, int pageNo, String filter) {
        if (language == null) {
            // singer cannot be null
            Log.d(TAG, "getHotSongsByLanguage.songsList is null.");
            return null;
        }

        int languageId = language.getId();
        String orderBy = "";  // no order by. Only the number that the song is ordered by descending order

        Retrofit localRetrofit = Client.getInstance();
        ApiInterface apiInterface = localRetrofit.create(ApiInterface.class);
        Call<SongList> call;
        if ( (filter == null) || (filter.isEmpty()) ) {
            call = apiInterface.getHotSongsByLanguageId(languageId, pageSize, pageNo);
        } else {
            call = apiInterface.getHotSongsByLanguageIdWithFilter(languageId, pageSize, pageNo, filter);
        }

        SongList songList = null;
        try {
            songList = call.execute().body();
        } catch (IOException ex) {
            ex.printStackTrace();
        }

        return songList;
    }

    public static SingerList getSingersBySingerType(SingerType singerType, int pageSize, int pageNo, String filter) {
        if (singerType == null) {
            // singerType cannot be null
            Log.d(TAG, "getSingersBySingerType.singersList is null.");
            return null;
        }

        int areaId = singerType.getId();
        String sex = singerType.getSex();
        String orderBy = "SingNa";  // singer's name

        Retrofit localRetrofit = Client.getInstance();
        ApiInterface apiInterface = localRetrofit.create(ApiInterface.class);
        Call<SingerList> call;
        if ( (filter == null) || (filter.isEmpty()) )  {
            call = apiInterface.getSingersBySingerTypeId(areaId, sex, pageSize, pageNo, orderBy);
        } else {
            call = apiInterface.getSingersBySingerTypeIdWithFilter(areaId, sex, pageSize, pageNo, orderBy, filter);
        }

        SingerList singerList = null;
        try {
            singerList = call.execute().body();
        } catch (IOException ex) {
            ex.printStackTrace();
        }

        return singerList;
    }

    public static SongList getSongsBySinger(Singer singer, int pageSize, int pageNo, String filter) {
        if (singer == null) {
            // singer cannot be null
            Log.d(TAG, "getSongsBySinger.songsList is null.");
            return null;
        }

        int singerId = singer.getId();
        // String orderBy = "NumWordsSongNa";  // order by (number of words + song's name)
        String orderBy = "SongNa";  // order by song's name

        Retrofit localRetrofit = Client.getInstance();
        ApiInterface apiInterface = localRetrofit.create(ApiInterface.class);
        Call<SongList> call;
        if ( (filter == null) || (filter.isEmpty()) )  {
            call = apiInterface.getSongsBySingerId(singerId, pageSize, pageNo, orderBy);
        } else {
            call = apiInterface.getSongsBySingerIdWithFilter(singerId, pageSize, pageNo, orderBy, filter);
        }

        SongList songList = null;
        try {
            songList = call.execute().body();
        } catch (IOException ex) {
            ex.printStackTrace();
        }

        return songList;
    }
}

File: app/src/main/java/com/smile/retrofit_package/RestApi.kt
package com.smile.retrofit_package

import android.util.Log
import com.smile.model.Language
import com.smile.model.LanguageList
import com.smile.model.Singer
import com.smile.model.SingerList
import com.smile.model.SingerType
import com.smile.model.SingerTypeList
import com.smile.model.SongList
import retrofit2.Callback

@JvmDefaultWithCompatibility
interface RestApi<T> : Callback<T> {

    companion object {
        const val TAG = "RestApi"
    }

    val callback : Callback<T>
        get() {
            return this
        }
    // get Retrofit client and Retrofit Api
    val apiInterface : ApiInterface
        get() {
            return Client.getInstance().create(ApiInterface::class.java)
        }

    @Suppress("UNCHECKED_CAST")
    fun getAllSingerTypes() {
        Log.d(TAG, "getAllSingerTypes")
        // get Call from Retrofit Api
        apiInterface.allSingerTypes.enqueue(callback as Callback<SingerTypeList>)
    }

    @Suppress("UNCHECKED_CAST")
    fun getAllLanguages() {
        Log.d(TAG, "getAllLanguages")
        // get Call from Retrofit Api
        apiInterface.allLanguages.enqueue(callback as Callback<LanguageList>)
    }

    @Suppress("UNCHECKED_CAST")
    fun getSongsBySinger(singer: Singer, pageSize: Int, pageNo: Int) {
        val singerId = singer.id
        val orderBy = "SongNa" // order by song's name
        apiInterface.getSongsBySingerId(singerId, pageSize, pageNo, orderBy)
            .enqueue(callback as Callback<SongList>)
    }

    @Suppress("UNCHECKED_CAST")
    fun getSongsBySinger(singer: Singer, pageSize: Int, pageNo: Int, filter: String) {
        val singerId = singer.id
        val orderBy = "SongNa" // order by song's name
        apiInterface.getSongsBySingerIdWithFilter(singerId, pageSize, pageNo, orderBy, filter)
            .enqueue(callback as Callback<SongList>)
    }

    @Suppress("UNCHECKED_CAST")
    fun getNewSongsByLanguage(language: Language, pageSize: Int, pageNo: Int) {
        Log.d(TAG, "getNewSongsByLanguage.no filter")
        val languageId = language.id
        // no order. Only the date that the song came in by descending order
        apiInterface.getNewSongsByLanguageId(languageId, pageSize, pageNo)
            .enqueue(callback as Callback<SongList>)
    }

    @Suppress("UNCHECKED_CAST")
    fun getNewSongsByLanguage(language: Language, pageSize: Int, pageNo: Int, filter: String) {
        Log.d(TAG, "getNewSongsByLanguage.filter not empty")
        val languageId = language.id
        // no order. Only the date that the song came in by descending order
        apiInterface.getNewSongsByLanguageIdWithFilter(languageId, pageSize, pageNo, filter)
            .enqueue(callback as Callback<SongList>)
    }

    @Suppress("UNCHECKED_CAST")
    fun getHotSongsByLanguage(language: Language, pageSize: Int, pageNo: Int) {
        Log.d(TAG, "getHotSongsByLanguage.no filter")
        val languageId = language.id
        // no order by. Only the number that the song is ordered by descending order
        apiInterface.getHotSongsByLanguageId(languageId, pageSize, pageNo)
            .enqueue(callback as Callback<SongList>)
    }

    @Suppress("UNCHECKED_CAST")
    fun getHotSongsByLanguage(language: Language, pageSize: Int, pageNo: Int, filter: String) {
        Log.d(TAG, "getHotSongsByLanguage.filter not empty")
        val languageId = language.id
        Log.d(TAG, "getNewSongsByLanguage.filter not empty.languageId = $languageId")
        Log.d(TAG, "getNewSongsByLanguage.filter not empty.pageSize = $pageSize")
        Log.d(TAG, "getNewSongsByLanguage.filter not empty.pageNo = $pageNo")
        Log.d(TAG, "getNewSongsByLanguage.filter not empty.filter = $filter")
        // no order by. Only the number that the song is ordered by descending order
        apiInterface.getHotSongsByLanguageIdWithFilter(languageId, pageSize, pageNo, filter)
            .enqueue(callback as Callback<SongList>)
    }

    @Suppress("UNCHECKED_CAST")
    fun getSongsByLanguage(language : Language, pageSize : Int, pageNo : Int) {
        Log.d(TAG, "getSongsByLanguage.no filter")
        val languageId = language.id
        // order by (number of words + song's name)
        val orderBy = "NumWordsSongNa"
        // get Call from Retrofit Api
        apiInterface.getSongsByLanguageIdOrderBy(languageId, pageSize, pageNo, orderBy)
            .enqueue(callback as Callback<SongList>)
    }

    @Suppress("UNCHECKED_CAST")
    fun getSongsByLanguage(language : Language, pageSize : Int, pageNo : Int, filter: String) {
        // filter cannot be empty
        Log.d(TAG, "getSongsByLanguage.filter not empty")
        val languageId = language.id
        // order by (number of words + song's name)
        val orderBy = "NumWordsSongNa"
        // get Call from Retrofit Api
        apiInterface.getSongsByLanguageIdOrderByWithFilter(
            languageId,
            pageSize,
            pageNo,
            orderBy,
            filter  // cannot be empty
        ).enqueue(callback as Callback<SongList>)
    }

    @Suppress("UNCHECKED_CAST")
    fun getSongsByLanguageNumOfWords(language: Language, numOfWords: Int, pageSize: Int
                                       , pageNo: Int) {
        Log.d(TAG, "getSongsByLanguageNumOfWords.no filter")
        val languageId = language.id
        val orderBy = "NumWordsSongNa" // order by (number of words + song's name)
        // get Call from Retrofit Api
        apiInterface.getSongsByLanguageIdNumOfWords(
            languageId,
            numOfWords,
            pageSize,
            pageNo,
            orderBy).enqueue(callback as Callback<SongList>)
    }

    @Suppress("UNCHECKED_CAST")
    fun getSongsByLanguageNumOfWords(language: Language, numOfWords: Int, pageSize: Int
                                     , pageNo: Int, filter: String) {
        Log.d(TAG, "getSongsByLanguageNumOfWords.filter not empty")
        val languageId = language.id
        val orderBy = "NumWordsSongNa" // order by (number of words + song's name)
        // get Call from Retrofit Api
        apiInterface.getSongsByLanguageIdNumOfWordsWithFilter(
            languageId,
            numOfWords,
            pageSize,
            pageNo,
            orderBy,
            filter).enqueue(callback as Callback<SongList>)
    }

    @Suppress("UNCHECKED_CAST")
    fun getSingersBySingerType(singerType: SingerType, pageSize: Int, pageNo: Int) {
        Log.d(TAG, "getSingersBySingerType.no filter")
        val areaId = singerType.id
        val sex = singerType.sex
        val orderBy = "SingNa" // singer's name
        apiInterface.getSingersBySingerTypeId(areaId, sex, pageSize,
            pageNo, orderBy).enqueue(callback as Callback<SingerList>)
    }

    @Suppress("UNCHECKED_CAST")
    fun getSingersBySingerType(singerType: SingerType, pageSize: Int, pageNo: Int, filter: String) {
        Log.d(TAG, "getSingersBySingerType.filter not empty")
        val areaId = singerType.id
        val sex = singerType.sex
        val orderBy = "SingNa" // singer's name
        apiInterface.getSingersBySingerTypeIdWithFilter(areaId, sex, pageSize,
            pageNo, orderBy, filter).enqueue(callback as Callback<SingerList>)
    }
}